azure-iot-device/tests/common/pipeline/pipeline_ops_test.py

```diff
@@ -0,0 +1,461 @@
+            )
+
+        @pytest.mark.it(
+            "Triggers and removes callbacks that have been added to the operation according to LIFO order"
+        )
+        def test_trigger_callbacks(self, mocker, cls_type, init_kwargs, error):
+            mocker.spy(handle_exceptions, "handle_background_exception")
+
+            # Set up callback mocks
+            cb1_mock = mocker.MagicMock()
+            cb2_mock = mocker.MagicMock()
+            cb3_mock = mocker.MagicMock()
+
+            def cb1(op, error):
+                # Callback 2 and 3 have already been triggered
+                assert cb1_mock.call_count == 1
+                assert cb2_mock.call_count == 1
+                assert cb3_mock.call_count == 1
+                assert len(op.callbacks) == 0
+
+            def cb2(op, error):
+                # Callback 3 has already been triggered, but Callback 1 has not
+                assert cb1_mock.call_count == 0
+                assert cb2_mock.call_count == 1
+                assert cb3_mock.call_count == 1
+                assert len(op.callbacks) == 1
+
+            def cb3(op, error):
+                # Callback 3 has been triggered, but no others have been.
+                assert cb1_mock.call_count == 0
+                assert cb2_mock.call_count == 0
+                assert cb3_mock.call_count == 1
+                assert len(op.callbacks) == 2
+
+            cb1_mock.side_effect = cb1
+            cb2_mock.side_effect = cb2
+            cb3_mock.side_effect = cb3
+
+            # Attach callbacks to op
+            init_kwargs["callback"] = cb1_mock
+            op = cls_type(**init_kwargs)
+            op.add_callback(cb2_mock)
+            op.add_callback(cb3_mock)
+            assert len(op.callbacks) == 3
+            assert not op.completed
+
+            # Run the completion
+            op.complete(error=error)
+
+            assert op.completed
+            assert cb3_mock.call_count == 1
+            assert cb3_mock.call_args == mocker.call(op=op, error=error)
+            assert cb2_mock.call_count == 1
+            assert cb2_mock.call_args == mocker.call(op=op, error=error)
+            assert cb1_mock.call_count == 1
+            assert cb1_mock.call_args == mocker.call(op=op, error=error)
+
+            # Because exceptions raised in callbacks are caught and sent to the background exception handler,
+            # the assertions in the above callbacks won't be able to directly raise AssertionErrors that will
+            # allow for testing normally. Instead we should check the background_exception_handler to see if
+            # any of the assertions raised errors and sent them there.
+            assert handle_exceptions.handle_background_exception.call_count == 0
+
+        @pytest.mark.it(
+            "Handles any Exceptions raised during execution of a callback by sending them to the background exception handler, and continuing on with completion"
```

Comment:
>continuing on with completion [](start = 131, length = 29)

I'm not convinced that continuing with completion is the correct behavior in this scenario.  Yes, we are reporting the exception (as best we can), but continuing with completion is also like ignoring the exception and that seems wrong.

Reply:
I'm open to changing the implementation, I'm not particularly tied to this logic.

---
In reply to: [346503501](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r346503501) [](ancestors = 346503501)

Reply:
I _think_ an exception in a callback becomes the op error.  This is a deviation from the old behavior, but I think your callback stack makes this better.

Say we have an op that has a stack with 3 callbacks.  The op completes with some status.  For our example, lets say it succeeds.

Now, inside the first callback, let's say there is some exception.   This means that something failed in the completion path for this op.  Can we say that this op succeeded?  No.  At this point, the best thing to do is to catch this exception and pass it as the op error to the next callback.

In a different comment thread, I talked about ops that change from error1->error2 or success->error.  This is how that might be accomplished and it feels perfectly intuitive to me.  

---
In reply to: [347047620](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r347047620) [](ancestors = 347047620,346503501)